feat: Add cpp_test folder with C++ test cases and CMake configuration

thedgarg31 opened the pull request:
**# What is the feature?**
* This PR introduces a new **`cpp_test` folder** under `tests/` to add C++ test cases.  
* It integrates the test suite into the existing CMake build system, ensuring that the tests are automatically included during the build process.  
* The feature improves the project’s modularity and enables better testing coverage.

**# How have you implemented the solution?** 
**Created a new `cpp_test` folder** under `tests/` with the following files:  
    - `main.cpp`: The entry point for running the C++ tests.  
    - `test_example.cpp`: Includes test cases to validate functionality.  
    - `CMakeLists.txt`: Configures the test suite and integrates it into the build process.  
**Modified the universal `CMakeLists.txt`:**
    - Included the new `cpp_test` folder to ensure it is compiled and executed with the project build.  
**Refactored and cleaned up the code**:
    - Improved the project structure by separating test cases into their own folder.
    - Verified the test execution locally.


 **# Does the PR impact any other area of the project, maybe another repo?**
 No, this PR does **not impact any other area** or external repositories.  
*The changes are **localized to the current project** and only affect the testing framework.

thedgarg31 commented:
Hello @kellijohnson-NOAA 
I’ve made the following updates in this PR:  

1. **Added a new `cpp_test` folder under `tests/`**, containing:  
   - `main.cpp`: Entry point for C++ tests.  
   - `test_example.cpp`: Test cases to validate functionality.  
   - `CMakeLists.txt`: CMake configuration for the new test suite.  

2. **Modified the universal `CMakeLists.txt`** to integrate the `cpp_test` folder into the build process.  

3. **Refactored and cleaned up the code** for better readability and efficiency.  

4. **Added new C++ test cases** and verified their correctness by running them locally.  

Please review the changes and let me know if you have any feedback or suggestions.  
Thank you for your time and guidance! 

Best regards,  
Daksh Garg

kellijohnson-NOAA commented:
@thedgarg31 I unfortunately do not see how your branch has changed? It still says that it is 1 commit ahead of main rather than one commit ahead of dev. I forgot to add in my instructions above that you will need to force push your branch up to your fork. So, the complete instructions are
```
git remote add upstream https://github.com/NOAA-FIMS/FIMS.git
git fetch upstream
git rebase upstream/dev
git status
```
Then you will need to fix the merge conflicts manually in the files that are listed as having conflicts and then force push, where the next set of instructions are just for the force push.
```
git push -f origin add-cpp-integration-tests
```
Another option is to just create a new branch off of upstream/dev, add your changes, and force push to this branch or push to a new branch. If you push to a new branch then you will have to open a new pull request.

thedgarg31 commented:
Hello @kellijohnson-NOAA 

**Documentation Update**
I have successfully added Doxygen documentation generation as per the proposal.  
Here’s what I’ve implemented:
- **CMake Integration:** Added Doxygen configuration in `CMakeLists.txt` to automatically generate HTML documentation during the build process.
- **Generated Documentation:** The HTML documentation is now available in the `build/docs/html` directory.
- **ZIP File Shared:** I have shared the ZIP file (`FIMS_Documentation.zip`) containing the complete HTML documentation with my mentor via email for review.

 **Next Steps:**
Please review the documentation and let me know if any modifications or improvements are required.

**Thanks for Your Support**
I truly appreciate your guidance and assistance in resolving previous issues and helping me implement this documentation successfully.

Kind Regards,
Daksh Garg

kellijohnson-NOAA commented:
Thanks. Did you mean to push all the files in latex/ and tests/ as well as FIMS_DOCUMENTATION.zip? I am guessing that those would remain local files and that we should add the repositories to our .gitignore file?

thedgarg31 commented:
You're right — the `latex/`, `tests/`, and `FIMS_DOCUMENTATION.zip` files are local artifacts generated by Doxygen and are not intended to be pushed to the repository. To prevent them from being included in future commits, I will add them to the `.gitignore` file.

kellijohnson-NOAA commented:
Can you please edit the last commit to exclude auxiliary files and folders so they are not tracked and edit the .gitignore so that they do not show up in future pull requests?

thedgarg31 commented:
"Hi @kellijohnson-NOAA, I have removed the auxiliary files and updated the .gitignore to prevent them from being tracked in future commits. The latest changes have been pushed. Please review and let me know if anything else needs to be adjusted. Thanks for your guidance!"

thedgarg31 commented:
No, I did not intend to remove the files inside the `cpp_test` folder — namely:

- `main.cpp`  
- `test_example.cpp`  
- `CMakeLists.txt`  

These files are essential components of the C++ test suite I contributed. The only files I intended to remove were the auto-generated documentation files and artifacts such as:

- `latex/`  
- `FIMS_DOCUMENTATION.zip`  
- any other non-source or build output folders mistakenly pushed earlier.

The `cpp_test` folder and its contents should remain in the repo as they are part of the new testing framework. Please let me know if any adjustments are needed.

kellijohnson-NOAA commented:
Thanks @thedgarg31 for the comment. You will need to add, commit, and push those files again because I cannot see them. You can navigate to [this link](https://github.com/NOAA-FIMS/FIMS/pull/784/files) to see what is and what is not included in this pull request.

thedgarg31 commented:
Thank you for your guidance and patience.

I have now re-added the cpp_test folder along with the essential files (main.cpp, CMakeLists.txt,) and pushed the updated changes to the branch. Kindly let me know if there’s anything further that needs to be addressed.

Looking forward to your feedback.
